File: WordPress/Classes/UIColor+Helpers.m
#import "UIColor+Helpers.h"
#import "NSString+HTML.h"

@implementation UIColor (Helpers)

// [UIColor UIColorFromRGBAColorWithRed:10 green:20 blue:30 alpha:0.8]
+ (UIColor *)UIColorFromRGBAColorWithRed: (CGFloat)r green:(CGFloat)g blue:(CGFloat)b alpha:(CGFloat)a {
    return [UIColor colorWithRed: r/255.0 green: g/255.0 blue: b/255.0 alpha:a];
}

// [UIColor UIColorFromRGBColorWithRed:10 green:20 blue:30]
+ (UIColor *)UIColorFromRGBColorWithRed:(CGFloat)r green:(CGFloat)g blue:(CGFloat)b {
    return [UIColor colorWithRed: r/255.0 green: g/255.0 blue: b/255.0 alpha: 0.5];
}

// [UIColor UIColorFromHex:0xc5c5c5 alpha:0.8];
+ (UIColor *)UIColorFromHex:(NSUInteger)rgb alpha:(CGFloat)alpha {
    return [UIColor colorWithRed:((float)((rgb & 0xFF0000) >> 16))/255.0
                          green:((float)((rgb & 0xFF00) >> 8))/255.0
                           blue:((float)(rgb & 0xFF))/255.0
                          alpha:alpha];
}

// [UIColor UIColorFromHex:0xc5c5c5];
+ (UIColor *)UIColorFromHex:(NSUInteger)rgb {
    return [UIColor UIColorFromHex:rgb alpha:1.0];
}

+ (UIColor *)colorWithHexString:(NSString *)hex {
    if ([hex length]!=6 && [hex length]!=3)
    {
        return nil;
    }
    
    NSUInteger digits = [hex length]/3;
    CGFloat maxValue = (digits==1)?15.0:255.0;
    
    CGFloat red = [[hex substringWithRange:NSMakeRange(0, digits)] integerValueFromHex]/maxValue;
    CGFloat green = [[hex substringWithRange:NSMakeRange(digits, digits)] integerValueFromHex]/maxValue;
    CGFloat blue = [[hex substringWithRange:NSMakeRange(2*digits, digits)] integerValueFromHex]/maxValue;
    
    return [UIColor colorWithRed:red green:green blue:blue alpha:1.0];
}

@end

File: Modules/Sources/WordPressUI/Extensions/UIColor+Helpers.swift
import UIKit

public extension UIColor {

    // A way to create dynamic colors that's compatible with iOS 11 & 12
    @objc
    convenience init(light: UIColor, dark: UIColor) {
        self.init { traitCollection in
            if traitCollection.userInterfaceStyle == .dark {
                return dark
            } else {
                return light
            }
        }
    }

    convenience init(color: UIColor) {
        var r: CGFloat = 0
        var g: CGFloat = 0
        var b: CGFloat = 0
        var a: CGFloat = 0

        color.getRed(&r, green: &g, blue: &b, alpha: &a)
        self.init(red: r, green: g, blue: b, alpha: a)
    }

    /// Creates a color based on a hexString. If the string is not a valid hexColor it return nil
    /// Example of colors: #FF0000, #00FF0000
    ///
    /// This method should only be used for parsing colors from third-party content  – if you're defining a color in code, use `fromHex`
    convenience init?(hexString: String) {
        let hex = hexString.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)

        guard let int = Scanner(string: hex).scanUInt64(representation: .hexadecimal) else {
            return nil
        }

        let a, r, g, b: UInt64
        switch hex.count {
        case 3: // RGB (12-bit)
            (a, r, g, b) = (255, (int >> 8) * 17, (int >> 4 & 0xF) * 17, (int & 0xF) * 17)
        case 6: // RGB (24-bit)
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB (32-bit)
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            return nil
        }
        self.init(red: CGFloat(r) / 255, green: CGFloat(g) / 255, blue: CGFloat(b) / 255, alpha: CGFloat(a) / 255)

    }

    convenience init(fromHex hex: UInt32) {
        self.init(
            red: CGFloat(UInt8((hex & 0xFF0000) >> 16)) / 0xFF,
            green: CGFloat(UInt8((hex & 0xFF00) >> 8)) / 0xFF,
            blue: CGFloat(UInt8(hex & 0xFF)) / 0xFF,
            alpha: 1.0
        )
    }

    convenience init(fromHex hex: UInt32, alpha: CGFloat) {
        self.init(
            red: CGFloat(UInt8((hex & 0xFF0000) >> 16)) / 0xFF,
            green: CGFloat(UInt8((hex & 0xFF00) >> 8)) / 0xFF,
            blue: CGFloat(UInt8(hex & 0xFF)) / 0xFF,
            alpha: alpha
        )
    }

    /// Produces an 6-digit CSS color hex string representation of `UIColor`.
    var hexString: String {
        guard let components = self.cgColor.components else {
            preconditionFailure("Color can't be represented in hexadecimal form")
        }

        if components.count == 2 {
            let w = Float(components[0])
            return String(format: "%02lX%02lX%02lX", lroundf(w * 255), lroundf(w * 255), lroundf(w * 255))
        }

        if components.count == 4 {
            let r = Float(components[0])
            let g = Float(components[1])
            let b = Float(components[2])

            return String(format: "%02lX%02lX%02lX", lroundf(r * 255), lroundf(g * 255), lroundf(b * 255))
        }

        preconditionFailure("Color can't be represented in hexadecimal form")
    }

    /// Produces an 8-digit CSS color hex string representation of `UIColor`.
    /// The last two digits represent the alpha value of the color.
    var hexStringWithAlpha: String {
        guard let components = self.cgColor.components else {
            preconditionFailure("Color can't be represented in hexadecimal form")
        }

        if components.count == 2 {
            let w = Float(components[0])
            let a = Float(components[1])

            return String(
                format: "%02lX%02lX%02lX%02lX",
                lroundf(w * 255),
                lroundf(w * 255),
                lroundf(w * 255),
                lroundf(a * 255)
            )
        }

        if components.count == 4 {
            let r = Float(components[0])
            let g = Float(components[1])
            let b = Float(components[2])
            let a = Float(components[3])

            return String(
                format: "%02lX%02lX%02lX%02lX",
                lroundf(r * 255),
                lroundf(g * 255),
                lroundf(b * 255),
                lroundf(a * 255)
            )
        }

        preconditionFailure("Color can't be represented in hexadecimal form")
    }

    // MARK: - Traits

    func color(for trait: UITraitCollection?) -> UIColor {
        if let trait {
            return resolvedColor(with: trait)
        }
        return self
    }

    func lightVariant() -> UIColor {
        return color(for: UITraitCollection(userInterfaceStyle: .light))
    }

    func darkVariant() -> UIColor {
        return color(for: UITraitCollection(userInterfaceStyle: .dark))
    }

    /// The same color with the dark and light variants swapped
    var variantInverted: UIColor {
        UIColor(light: darkVariant(), dark: lightVariant())
    }
}
